actix-http/src/h1/dispatcher.rs

```diff
@@ -645,83 +645,82 @@ where
         cx: &mut Context<'_>,
     ) -> Result<(), DispatchError> {
         let mut this = self.as_mut().project();
-        if this.ka_timer.is_none() {
-            // shutdown timeout
-            if this.flags.contains(Flags::SHUTDOWN) {
-                if let Some(interval) = this.codec.config().client_disconnect_timer() {
-                    this.ka_timer.set(Some(sleep_until(interval)));
-                } else {
-                    this.flags.insert(Flags::READ_DISCONNECT);
-                    if let Some(mut payload) = this.payload.take() {
-                        payload.set_error(PayloadError::Incomplete(None));
+
+        // when a branch is not explicit return early it's meant to fall through
+        // and return as Ok(())
+        match this.ka_timer.as_mut().as_pin_mut() {
+            None => {
+                // conditionally go into shutdown timeout
+                if this.flags.contains(Flags::SHUTDOWN) {
+                    if let Some(deadline) = this.codec.config().client_disconnect_timer()
+                    {
+                        // write client disconnect time out and poll again to
+                        // go into Some<Pin<&mut Sleep>> branch
+                        this.ka_timer.set(Some(sleep_until(deadline)));
+                        return self.poll_keepalive(cx);
```

Comment:
why recurse? we wouldnt expect the keepalive to proc instantly

Reply:
timer has to be polled once for registering waker. It's either a short recurse to make it happen or write this
```rust
this.ka_timer.as_mut().set(Some(sleep_until(deadline)));
let _ = this.ka_timer.as_mut().as_pin_mut().unwrap().poll(cx);
```

Reply:
true, and it's not on the hot path since processing is paused on socket idle, i agree with your choice